File: app/src/main/java/com/shuyu/apprecycler/chat/detail/ChatDetailActivity.java
package com.shuyu.apprecycler.chat.detail;

import android.content.Context;
import android.graphics.Color;
import android.os.Bundle;
import android.support.annotation.Nullable;
import android.support.v7.app.AppCompatActivity;
import android.support.v7.widget.LinearLayoutManager;
import android.support.v7.widget.RecyclerView;
import android.support.v7.widget.Toolbar;
import android.view.MotionEvent;
import android.view.View;

import com.shuyu.apprecycler.R;
import com.shuyu.apprecycler.chat.detail.dagger.ChatDetailAdapter;
import com.shuyu.apprecycler.chat.detail.dagger.ChatDetailViewControl;
import com.shuyu.apprecycler.chat.detail.dagger.ChatDetailViewOption;
import com.shuyu.apprecycler.chat.detail.dagger.ChatSuperAdapterManager;
import com.shuyu.apprecycler.chat.detail.dagger.component.DaggerChatDetailComponent;
import com.shuyu.apprecycler.chat.detail.dagger.module.ChatDetailManagerModule;
import com.shuyu.apprecycler.chat.detail.dagger.module.ChatDetailPresenterModule;
import com.shuyu.apprecycler.chat.detail.dagger.module.ChatDetailViewModule;
import com.shuyu.apprecycler.chat.detail.view.ChatDetailBottomView;
import com.shuyu.apprecycler.chat.detail.view.ChatDetailEmojiLayout;
import com.shuyu.bind.listener.OnItemClickListener;
import com.shuyu.bind.listener.OnLoadingListener;
import com.shuyu.textutillib.EmojiLayout;
import com.shuyu.textutillib.KeyBoardLockLayout;
import com.shuyu.textutillib.RichEditText;

import javax.inject.Inject;

import butterknife.BindView;
import butterknife.ButterKnife;
import butterknife.OnClick;
import butterknife.OnFocusChange;

/**
 * 聊天详情
 * Created by guoshuyu on 2017/9/4.
 */

public class ChatDetailActivity extends AppCompatActivity implements ChatDetailContract.IChatDetailView, OnItemClickListener, OnLoadingListener, View.OnTouchListener {

    @BindView(R.id.chat_detail_activity_toolbar)
    Toolbar mChatDetailActivityToolbar;
    @BindView(R.id.chat_detail_activity_recycler)
    RecyclerView mChatDetailActivityRecycler;
    @BindView(R.id.chat_detail_activity_edit)
    RichEditText mChatDetailActivityEdit;
    @BindView(R.id.chat_detail_activity_keyboard_layout)
    KeyBoardLockLayout mChatDetailActivityKeyboardLayout;
    @BindView(R.id.chat_detail_activity_send_emojiLayout)
    EmojiLayout mChatDetailActivitySendEmojiLayout;
    @BindView(R.id.chat_detail_activity_bottom_menu)
    ChatDetailBottomView mChatDetailActivityBottomMenu;

    @Inject
    ChatDetailPresenter mPresenter;
    @Inject
    ChatSuperAdapterManager mNormalAdapterManager;
    @Inject
    ChatDetailAdapter mAdapter;
    @Inject
    ChatDetailViewControl mChatDetailViewControl;

    @Override
    protected void onCreate(@Nullable Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        ChatDetailEmojiLayout.initEmoji(getContext());
        setContentView(R.layout.activity_chat_detail);
        ButterKnife.bind(this);
        initTitle();
        initActivity();
    }

    private void initTitle() {
        mChatDetailActivityToolbar.setTitle("聊天详情");
        mChatDetailActivityToolbar.setTitleTextColor(Color.WHITE);
        setSupportActionBar(mChatDetailActivityToolbar);
    }

    @SuppressWarnings("unchecked")
    private void initActivity() {

        ChatDetailViewOption option = new ChatDetailViewOption();
        option.setChatDetailActivityBottomMenu(mChatDetailActivityBottomMenu)
                .setChatDetailActivityKeyboardLayout(mChatDetailActivityKeyboardLayout)
                .setChatDetailActivitySendEmojiLayout(mChatDetailActivitySendEmojiLayout)
                .setChatDetailActivityEdit(mChatDetailActivityEdit)
                .setRecyclerView(mChatDetailActivityRecycler);

        DaggerChatDetailComponent.builder()
                .chatDetailPresenterModule(new ChatDetailPresenterModule(this))
                .chatDetailManagerModule(new ChatDetailManagerModule(this, this, this))
                .chatDetailViewModule(new ChatDetailViewModule(option))
                .build()
                .inject(this);
    }

    @OnClick(R.id.chat_detail_activity_send)
    public void onSendClicked() {
        mPresenter.sendMsg(mChatDetailActivityEdit.getText().toString());
    }

    @OnClick(R.id.chat_detail_activity_edit)
    public void onEditClicked() {
        mChatDetailViewControl.showKeyBoradOnly();
    }

    @OnClick(R.id.chat_detail_activity_emoji_btn)
    public void onEmojiBtnClicked() {
        mChatDetailViewControl.showEmojiOnly();
    }

    @OnClick(R.id.chat_detail_activity_bottom_btn)
    public void onBottomViewBtnClicked() {
        mChatDetailViewControl.showBottomMenuOnly();
    }

    @OnFocusChange(R.id.chat_detail_activity_edit)
    public void onEditFocusChanged(boolean hasFocus) {
        if (hasFocus) {
            mChatDetailViewControl.showKeyBoradOnly();
        }
    }

    @Override
    public Context getContext() {
        return this;
    }

    //TODO 用diff判断
    @Override
    public void notifyView() {
        mAdapter.notifyDataSetChanged();
    }

    @Override
    public void sendSuccess() {
        mChatDetailActivityEdit.setText("");
        mAdapter.notifyDataSetChanged();
    }

    @Override
    public void onRefresh() {

    }

    @Override
    public void onLoadMore() {

    }

    @Override
    public void onItemClick(Context context, int position) {

    }

    @Override
    public boolean onTouch(View v, MotionEvent event) {
        if (v instanceof RecyclerView) {
            mChatDetailViewControl.hideAllMenuAndKebBoard();
        }
        return false;
    }

    @Override
    public void onBackPressed() {
        if (mChatDetailViewControl.interceptBack()) {
            return;
        }
        super.onBackPressed();
    }
}

File: app-kotlin/src/main/java/example/shuyu/recycler/kotlin/chat/detail/ChatDetailActivity.kt
package example.shuyu.recycler.kotlin.chat.detail

import android.content.Context
import android.graphics.Color
import android.os.Bundle
import android.support.v7.app.AppCompatActivity
import android.support.v7.widget.RecyclerView
import android.view.MotionEvent
import android.view.View

import com.shuyu.commonrecycler.listener.OnItemClickListener
import com.shuyu.commonrecycler.listener.OnLoadingListener
import com.shuyu.textutillib.EmojiLayout

import javax.inject.Inject

import example.shuyu.recycler.kotlin.R
import example.shuyu.recycler.kotlin.chat.detail.dagger.ChatDetailAdapter
import example.shuyu.recycler.kotlin.chat.detail.dagger.ChatDetailViewControl
import example.shuyu.recycler.kotlin.chat.detail.dagger.ChatDetailViewOption
import example.shuyu.recycler.kotlin.chat.detail.dagger.ChatSuperAdapterManager
import example.shuyu.recycler.kotlin.chat.detail.dagger.component.DaggerChatDetailComponent
import example.shuyu.recycler.kotlin.chat.detail.dagger.module.ChatDetailManagerModule
import example.shuyu.recycler.kotlin.chat.detail.dagger.module.ChatDetailPresenterModule
import example.shuyu.recycler.kotlin.chat.detail.dagger.module.ChatDetailViewModule
import example.shuyu.recycler.kotlin.chat.detail.view.ChatDetailBottomView
import kotlinx.android.synthetic.main.activity_chat_detail.*


/**
 * 聊天详情
 * Created by guoshuyu on 2017/9/4.
 */

class ChatDetailActivity : AppCompatActivity(), ChatDetailContract.IChatDetailView, OnItemClickListener, OnLoadingListener, View.OnTouchListener, View.OnClickListener {

    @Inject
    lateinit var mPresenter: ChatDetailPresenter
    @Inject
    lateinit var mNormalAdapterManager: ChatSuperAdapterManager
    @Inject
    lateinit var mAdapter: ChatDetailAdapter
    @Inject
    lateinit var mChatDetailViewControl: ChatDetailViewControl

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_chat_detail)
        initTitle()
        initActivity()
        initListener()
    }

    private fun initTitle() {
        chat_detail_activity_toolbar?.title = "聊天详情"
        chat_detail_activity_toolbar?.setTitleTextColor(Color.WHITE)
        setSupportActionBar(chat_detail_activity_toolbar)
    }

    private fun initActivity() {

        val option = ChatDetailViewOption()
        option.setChatDetailActivityBottomMenu(chat_detail_activity_bottom_menu as ChatDetailBottomView)
                .setChatDetailActivityKeyboardLayout(chat_detail_activity_keyboard_layout)
                .setChatDetailActivitySendEmojiLayout(chat_detail_activity_send_emojiLayout as EmojiLayout)
                .setChatDetailActivityEdit(chat_detail_activity_edit)
                .setRecyclerViewOption(chat_detail_activity_recycler)

        DaggerChatDetailComponent.builder()
                .chatDetailPresenterModule(ChatDetailPresenterModule(this))
                .chatDetailManagerModule(ChatDetailManagerModule(this, this, this))
                .chatDetailViewModule(ChatDetailViewModule(option))
                .build()
                .inject(this)
    }

    private fun initListener() {
        chat_detail_activity_send.setOnClickListener(this)
        chat_detail_activity_edit.setOnClickListener(this)
        chat_detail_activity_emoji_btn.setOnClickListener(this)
        chat_detail_activity_bottom_btn.setOnClickListener(this)
        chat_detail_activity_edit.setOnFocusChangeListener { _, _ -> mChatDetailViewControl.showKeyBoradOnly() }
    }

    override fun onClick(v: View?) {
        when (v?.id) {
            R.id.chat_detail_activity_send -> {
                mPresenter.sendMsg(chat_detail_activity_edit?.text.toString())
            }
            R.id.chat_detail_activity_edit -> {
                mChatDetailViewControl.showKeyBoradOnly()
            }
            R.id.chat_detail_activity_emoji_btn -> {
                mChatDetailViewControl.showEmojiOnly()
            }
            R.id.chat_detail_activity_bottom_btn -> {
                mChatDetailViewControl.showBottomMenuOnly()
            }
        }
    }

    override fun onItemClick(context: Context, position: Int) {

    }

    override fun onTouch(v: View, event: MotionEvent): Boolean {
        if (v is RecyclerView) {
            mChatDetailViewControl.hideAllMenuAndKebBoard()
        }
        return false
    }

    override fun onBackPressed() {
        if (mChatDetailViewControl.interceptBack()) {
            return
        }
        super.onBackPressed()
    }

    override fun onDestroy() {
        super.onDestroy()
        mPresenter.release()
    }

    override fun onRefresh() {

    }

    override fun onLoadMore() {
        mPresenter.loadMoreData()
    }

    override fun notifyView() {
        mAdapter.notifyDataSetChanged()
    }


    override fun loadMoreComplete() {
        mNormalAdapterManager.loadMoreComplete()
    }

    override fun loadMoreEnd() {

        mNormalAdapterManager.setNoMore(true)
    }

    override fun sendSuccess() {
        chat_detail_activity_edit?.setText("")
        mAdapter.notifyDataSetChanged()
    }

    override val context: Context
        get() = this
}